File: leetcode/python/520_detect_capital.py
import re

class Solution:
    def detectCapitalUse(self, word: str) -> bool:
        if len(word) < 2:
            return True
        if word[0].isupper() and word[1].isupper():
            return self.isAllUppercase(word[2:])
        if word[1].islower():
            return self.isAllLowercase(word[2:])
        return False
    
    def isAllUppercase(self, s: str) -> str:
        for c in s:
            if c.islower():
                return False
        return True

    def isAllLowercase(self, s: str) -> str:
        for c in s:
            if c.isupper():
                return False
        return True

    def detectCapitalUse(self, word: str) -> bool:
        return re.fullmatch(r"[A-Z]*|.[a-z]*", word)

s = Solution()
print(s.detectCapitalUse('USA'))
print(s.detectCapitalUse('FlaG'))
print(s.detectCapitalUse('leetcode'))
print(s.detectCapitalUse('Google'))

File: leetcode/c++/520_detect_capital.cpp
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
    bool detectCapitalUse(string word) {
        auto n {word.size()};
        if (n==1) return true;

        int cnt {0};
        for (auto c: word) {
            if (isupper(c)) cnt++;
        }
        if (cnt==0 || cnt==n) return true;
        if (cnt==1 && isupper(word[0])) return true;
        return false;
    }
};